pkg/apis/options/login_url_parameters.go

```diff
@@ -0,0 +1,54 @@
+package options
+
+import "regexp"
+
+// LoginURLParameter is the configuration for a single query parameter that
+// can be passed through from the `/oauth2/start` endpoint to the IdP login
+// URL.  The "default" option specifies the default value or values (if any)
+// that will be passed to the IdP for this parameter, and "allowed" is a
+// regular expression specifying which values (if any) that a caller passes
+// in the query string to `/oauth2/start` are permitted to override the defaults.
+// If _only_ a default is specified and no "allowed" then the parameter is
+// effectively fixed - the default value will always be used and anything
+// passed to the start URL will be ignored.  If _only_ "allowed" is specified
+// but no default then the parameter will only be passed on to the IdP if
+// the caller provides it, and no value will be sent otherwise.  Note that
+// the "allowed" regex is implicitly anchored with a `^` at the start and `$`
+// at the end, so it must match the _whole_ of the value to be considered
+// valid.
+type LoginURLParameter struct {
+	// Default specifies a default value or values that will be
+	// passed to the IdP if not overridden.
+	Default []string `json:"default,omitempty"`
+
+	// Allowed specifies a regular expression to validate any
+	// override values passed to the start endpoint.
+	Allowed *URLParameterRegexp `json:"allowed,omitempty"`
+}
+
+type URLParameterRegexp struct {
+	re *regexp.Regexp
+}
+
+func (r *URLParameterRegexp) UnmarshalText(b []byte) error {
+	regex, err := regexp.Compile("^(?:" + string(b) + ")$")
+	if err != nil {
+		return err
+	}
+	r.re = regex
+
+	return nil
+}
+
+func (r *URLParameterRegexp) MarshalText() ([]byte, error) {
+	if r.re != nil {
+		s := r.re.String()
+		// strip off the ^(?:...)$ wrapper
+		return []byte(s[4 : len(s)-2]), nil
+	}
+	return nil, nil
+}
+
+func (r URLParameterRegexp) MatchString(s string) bool {
+	return r.re.MatchString(s)
+}
```

Comment:
Please add doc strings to these methods and in particular, note which interface these are implementing